### PR TITLE
Feat/click event fanout

### PR DESCRIPTION
**Pull request by Zingzy**
## Summary by Sourcery

Introduce a click event fanout pipeline that emits immutable ClickEvent objects from redirect routes into a Redis Stream, processes them asynchronously via a dedicated click worker, and falls back to inline tracking when the stream infrastructure is unavailable or misconfigured.

New Features:
- Add a ClickEvent DTO and event sink abstraction so redirect routes emit structured click events instead of calling the click service directly.
- Introduce a Redis Stream–backed click sink and a FastStream-based click worker that consumes click events for stats and hot-URL detection, with configurable consumer groups and dead-lettering.
- Add optional hotness detection for high-traffic short URLs powered by Redis windowed counters and pluggable actions.
- Expose ClickEventsSettings in application config to control sink mode, stream/Redis settings, and worker behavior, with safe inline defaults.
- Wire a dedicated queue Redis and click worker services into local and production Docker configurations, gated behind opt-in env and compose profiles.

Enhancements:
- Refactor redirect routes to snapshot resolved URL data into ClickEvents, strip sensitive password hashes, and perform pre-emit bot blocking for legacy schemas while keeping v2 bot handling in the pipeline.
- Centralize ClickService construction for reuse between the web app and the click worker to ensure schema handler parity.
- Improve Redis client creation with richer connection options, labeled logging, and support for multiple Redis roles (cache vs click queue).

Build:
- Add FastStream (Redis) to project dependencies to support the click worker process.

Deployment:
- Extend docker-compose (dev and prod) with a dedicated queue Redis and click-worker service, including health checks, memory policies, and profiles for opt-in deployment of the click event pipeline.

Tests:
- Expand integration tests around redirect and click tracking to assert on emitted ClickEvents, bot-block behavior, HEAD/password edge cases, and resilience to sink errors.
- Add unit tests for click event encoding/decoding, click sinks, stats and hotness consumers, DLQ behavior, click events settings, and click worker wiring to validate the new pipeline end-to-end.

## Summary by CodeRabbit

* **New Features**
  * Added optional click-event streaming support for redirects, including background processing and Redis-backed queueing.
  * Introduced a worker that processes click analytics and hot-traffic detection in the background.
  * Added health and telemetry checks for the new click-processing pipeline.

* **Bug Fixes**
  * Improved redirect behavior for bot traffic, password-protected links, and HEAD requests.
  * Ensured click tracking failures do not interrupt normal redirects in most cases.
  * Improved cleanup and fallback behavior when Redis connections or queue processing are unavailable.

### CODE REVIEW

File: docker-compose.prod.yml [hunk 1]
```diff
@@ -209,10 +209,88 @@ services:
...
+      --maxmemory-policy noeviction
+      --appendonly yes
+      --appendfsync everysec
+      --tcp-keepalive 60
+      --timeout 300
+      --requirepass ${REDIS_QUEUE_PASSWORD}
+    volumes:
+      - redis-queue-data:/data
+    healthcheck:
+      test: ["CMD", "redis-cli", "ping"]
```
[the remark this answers is not in the document]
Comment (Zingzy) [reply docker-compose.prod.yml]: Fixed in 86a46d8. One correction: the failure mode was false-healthy, not unhealthy — redis-cli exits 0 on a NOAUTH reply, so the check silently degraded to a TCP probe. Both redis services now auth via REDISCLI_AUTH and grep for PONG.

File: workers/click_worker.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    runtime = _WorkerRuntime(
+        mongo_client=mongo_client,
+        cache_redis=cache_redis,
+        counter_redis=None,
+    )
+
+    if "stats" in groups:
+        geoip = GeoIPService(settings.geoip_country_db, settings.geoip_city_db)
+        url_cache = UrlCache(cache_redis, ttl_seconds=settings.redis.redis_ttl_seconds)
+        runtime.consumers["stats"] = StatsClickConsumer(
```
[the remark this answers is not in the document]
Comment (Zingzy) [reply workers/click_worker.py:135]: Not applicable — GeoIPService has no close/aclose API, and the main app intentionally never tears it down either. It's a process-lifetime singleton; the mmdb readers are reclaimed at process exit.

File: workers/dlq.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        try:
+            await redis.xadd(
+                self._dlq_stream,
+                {
+                    "__data__": json.dumps(payload, default=str),
+                    DLQ_FIELD_SOURCE_ID: message_id,
+                    DLQ_FIELD_GROUP: self._group,
+                    DLQ_FIELD_REASON: _REASON_MAX_DELIVERIES,
+                },
+            )
```
[the remark this answers is not in the document]
Comment (Zingzy) [reply workers/dlq.py:73]: Fixed in 86a46d8 — the DLQ XADD is now bounded with MAXLEN ~10k.

File: workers/dlq.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                group=self._group,
+                error=str(exc),
+                error_type=type(exc).__name__,
+            )
+            return None
+        if not pending:
+            return None
+        entry = pending[0]
+        times = entry.get("times_delivered")
+        return int(times) if times is not None else None
```
[the remark this answers is not in the document]
Comment (Zingzy) [reply workers/dlq.py:115]: Fixed in 86a46d8 — removed. Leftover from before decode_responses=True made byte coercion unnecessary.

File: config.py [hunk 5]
```diff
@@ -185,6 +186,77 @@ def _validate_cf_saas_config(self) -> CustomDomainSettings:
...
+    # Consumer tunables.
+    batch_size: int = Field(default=100, ge=1)
+    block_ms: int = Field(default=2000, ge=100)
+    # Pending messages idle longer than claim_idle_ms are reclaimed by the
+    # claimer subscriber (XAUTOCLAIM); after max_deliveries failed attempts
+    # a message is dead-lettered to dlq_stream.
+    claim_idle_ms: int = Field(default=60_000, ge=1000)
+    max_deliveries: int = Field(default=5, ge=1)
+    stats_interval_seconds: float = Field(default=30.0, gt=0)
+
```
[the remark this answers is not in the document]
Comment (Zingzy) [reply config.py:238]: Stale — it's wired into StreamMetricsReporter (workers/click_worker.py:172) as of 63427e2.

File: infrastructure/cache/redis_client.py [hunk 6]
```diff
@@ -12,18 +12,39 @@
...
         )
         return None
     except Exception as e:
-        log.warning("redis_unexpected_error", error=str(e), error_type=type(e).__name__)
+        log.warning(
+            "redis_unexpected_error",
+            label=label,
+            error=str(e),
+            error_type=type(e).__name__,
+        )
```
[the remark this answers is not in the document]
Comment (Zingzy) [reply infrastructure/cache/redis_client.py:56]: Fixed in 86a46d8 — the client is now closed on construction/ping failure before returning None.

File: services/click/events.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    raw = fields.get(STREAM_FIELD_DATA)
+    if not raw:
+        log.warning("click_event_missing_data", fields=list(fields.keys()))
+        return None
+    try:
+        return ClickEvent.model_validate_json(raw)
+    except PydanticValidationError:
+        log.warning("click_event_malformed", raw=raw[:300])
+        return None
+
```
[the remark this answers is not in the document]
Comment (Zingzy) [reply services/click/events.py:107]: Fixed in 86a46d8 — malformed-event logs now carry only pydantic error locs/types (include_input=False); the raw payload no longer reaches logs.

File: services/click/sinks/stream.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        self._fallback = fallback
+
+    async def emit(self, event: ClickEvent) -> None:
+        try:
+            await self._redis.xadd(
+                self._stream,
+                to_stream_fields(event),
+                maxlen=self._maxlen,
+                approximate=True,
+            )
```
[the remark this answers is not in the document]
Comment (Zingzy) [reply services/click/sinks/stream.py:48]: Fixed in 86a46d8 — create_redis_client now sets socket_timeout/socket_connect_timeout (5s), so a wedged connection raises and the sink falls back inline instead of stalling redirects. Good catch.

File: workers/click_worker.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+    async def aclose(self) -> None:
+        for task in self.telemetry_tasks:
+            task.cancel()
+        await self.mongo_client.close()
+        if self.cache_redis is not None:
+            await self.cache_redis.aclose()
+        if self.counter_redis is not None:
+            await self.counter_redis.aclose()
+        if self.telemetry_redis is not None:
+            await self.telemetry_redis.aclose()
```
[the remark this answers is not in the document]
Comment (Zingzy) [reply workers/click_worker.py:97]: Fixed in 86a46d8 — telemetry tasks are now gathered after cancel() before their redis client is closed.
